### PR TITLE
GH-111485: Separate out parsing, analysis and code-gen phases

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-111485

### CODE REVIEW

File: Tools/cases_generator/tier1_generator.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+
+class Stack:
+
+    def __init__(self):
+        self.top_offset: StackOffset = StackOffset()
+        self.base_offset: StackOffset = StackOffset()
+        self.peek_offset: StackOffset = StackOffset()
+        self.variables: list[StackItem] = []
+        self.defined: Set[str] = set()
```
Comment (gvanrossum) [inline Tools/cases_generator/tier1_generator.py]: Set -> set.

File: Tools/cases_generator/tier1_generator.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+from cwriter import CWriter
+from typing import Tuple, TextIO, Iterator
+from lexer import Token
+from dataclasses import dataclass
+from stack import StackOffset
+
+def write_header(filename: str, outfile: TextIO) -> None:
+    outfile.write(f"""// This file is generated by {__file__}
+// from:
+//   {filename}
```
Comment (gvanrossum) [inline Tools/cases_generator/tier1_generator.py]: Losing some subtlety here -- the original code prints the same filenames (both for the script and the input file) regardless of what the current directory is or the platform.

File: Tools/cases_generator/cwriter.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
+
+from lexer import Token
+
+class CWriter:
+    'A writer that understands how to format C code.'
+
+    def __init__(self, out, indent = 0):
```
Comment (gvanrossum) [inline Tools/cases_generator/cwriter.py]: I would like to get the type annotations back. They help understanding the code.

File: Tools/cases_generator/tier1_generator.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        print(__file__, "INFILE", "OUTFILE")
+    else:
+        filename = sys.argv[1]
+        data = analyze_file(filename)
+        if len(sys.argv) == 3:
+            outfilename = sys.argv[2]
+            with open(outfilename, "w") as outfile:
+                generate_tier1(filename, data, outfile)
+        else:
+            generate_tier1(filename, data, sys.stdout)
```
Comment (gvanrossum) [inline Tools/cases_generator/tier1_generator.py]: Honestly we can drop this part and just insist on an output filename. I've never wanted to use this since the generator went into production.

File: Tools/cases_generator/lexer.py [hunk 5]
```diff
@@ -112,7 +112,7 @@ def choice(*opts: str) -> str:
 char = r"\'.\'"  # TODO: escape sequence
 CHARACTER = "CHARACTER"
 
-comment_re = r"//.*|/\*([^*]|\*[^/])*\*/"
+comment_re = r"(//.*\n)|/\*([^*]|\*[^/])*\*/"
```
Comment (gvanrossum) [inline Tools/cases_generator/lexer.py]: I'm not excited that the `\n` is now considered part of the comment. It just seems to uglify other things (like the begin/end markers) and the lexer will need to skip newlines in other contexts anyways.

File: Tools/cases_generator/cwriter.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    def emit_token(self, tkn):
+        if self.line < tkn.line:
+            self.line = tkn.line
+            self.newline = True
+            self.out.write("\n")
+        if self.newline:
+            self.column = 0
+        elif self.column > 0 and self.column < tkn.column:
+            self.out.write(" ") * (tkn.column-self.column)
```
Comment (gvanrossum) [inline Tools/cases_generator/cwriter.py]: I think this is responsible for blank lines containing lots of spaces (up to the indent level) in the generated file.

File: Tools/cases_generator/cwriter.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        self.line = tkn.end_line
+        if tkn.kind == "COMMENT" and "\n" in tkn.text:
+            # Try to format multi-line comments sensibly
```
Comment (gvanrossum) [inline Tools/cases_generator/cwriter.py]: This seems a near miss so far -- comments like this
```c
/* blah
 * blah blah
 */
```
end up looking line this:
```c
/* blah
   * blah blah
   */
```

File: Tools/cases_generator/cwriter.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+
```
Comment (gvanrossum) [inline Tools/cases_generator/cwriter.py]: I don't know where it's happening, but I found at least one example where
```
if (err != 0) {
```
is not indented. (L2183 in the new output.)

There's also an unindented block in `YIELD_VALUE` around L5816.

File: Tools/cases_generator/analyzer.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+              ", jumps:", self.jumps,
+              ", eval_breaker:", self.ends_with_eval_breaker,
+              ", needs_this:", self.needs_this,
+              ", always_exits:", self.always_exits,
+              ", stores_sp:", self.stores_sp,
+              sep=""
+        )
+
+
+SKIP_PROPERTIES = Properties(
```
Comment (gvanrossum) [inline Tools/cases_generator/analyzer.py:39]: I'm sure dataclasses let you use keyword args and defaults here. :-)

File: Tools/cases_generator/analyzer.py [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+class PseudoInstruction:
+    name: str
+    targets: list[Instruction]
+
+    def dump(self, indent):
+        print(indent, self.name, "->", " or ".join([t.name for t in self.targets]))
+
+
+@dataclass
+class Family:
```
Comment (gvanrossum) [inline Tools/cases_generator/analyzer.py:167]: This reminds me of another possible naming scheme. We already have `parser.InstDef` vs. `Instruction`. Maybe the other top-level things that the parser recognizes could also be named `FooDef`? I.e., `FamilyDef`, `PseudoDef`, `MacroDef`.

**Review by gvanrossum — COMMENTED** [review]
It's more of a rewrite (or everything besides the parser and lexer) than a refactor. :-) It does feel cleaner, so I support forging ahead here.

General: I'd like it to use argparse, have correct type annotations everywhere, and be formatted using Black.

I'd like to understand why `_PUSH_FRAME` and the macros using it are now considered not to have an output stack effect (or one less, for the macros). Given that the generated output is unchanged for these, I am assuming that you're no longer checking that all the members of a family have the same stack effect?

One concern I have with creating separate top-level command-line utilities is that in my experience, the bulk of the processing time is in the parser -- which means that if you have five utilities to generate each of the five output files, you end up waiting five times as long for `make regen-cases` or its equivalent. This should be easy to address given your architecture. Please don't forget to update README.md.

I didn't try to understand your stack.py carefully, but it deserves a lot of scrutiny. In the past fixes there usually were obvious when the generated code was different -- but you've moved things around in the output just enough (even taking the case-sorting out of the equation) that it's hard to review everything without glazing over.

File: Tools/cases_generator/analyzer.py (hunk 10 above)
Comment (markshannon) [reply Tools/cases_generator/analyzer.py:167]: That sounds like a good scheme.

**Comment by markshannon** [comment]
> One concern I have with creating separate top-level command-line utilities is that in my experience, the bulk of the processing time is in the parser -- which means that if you have five utilities to generate each of the five output files, you end up waiting five times as long for make regen-cases or its equivalent.

With separate tools we can have separate dependencies for each generated file. No need for `make regen-cases` at all, just the normal `make` which will run the jobs in parallel. 

It is the lexer that takes most of the time. If that's an issue, we can make it faster.

**Comment by gvanrossum** [comment]
> With separate tools we can have separate dependencies for each generated file. No need for `make regen-cases` at all, just the normal `make` which will run the jobs in parallel.

Great, I'd  like implicit dependencies rather than the need to make regen-case.

> It is the lexer that takes most of the time. If that's an issue, we can make it faster.

Maybe through caching the analysis? Looks picklable.

**Comment by gvanrossum** [comment]
> > It is the lexer that takes most of the time. If that's an issue, we can make it faster.
> 
> Maybe through caching the analysis? Looks picklable.

I just realized I don't even care. `make -j` will run them all in parallel.

File: Makefile.pre.in [hunk 11]
```diff
@@ -1587,14 +1587,15 @@ regen-cases:
 	$(PYTHON_FOR_REGEN) \
 	    $(srcdir)/Tools/cases_generator/generate_cases.py \
 		$(CASESFLAG) \
-		-o $(srcdir)/Python/generated_cases.c.h.new \
```
Comment (gvanrossum) [inline Makefile.pre.in:1590]: I'd put this line back; the default always overwrites Python/generated_cases.c.h.

File: Python/generated_cases.c.h [hunk 12]
```diff
@@ -4625,29 +4664,31 @@
...
                 break;
-            default:
+                default:
                 _PyErr_SetString(tstate, PyExc_SystemError,
                                  "bad RAISE_VARARGS oparg");
                 break;
             }
-            if (true) { STACK_SHRINK(oparg); goto error; }
+            if (true) { stack_pointer += -oparg; goto error; }
+            stack_pointer += -oparg;
```
Comment (gvanrossum) [inline Python/generated_cases.c.h]: I get a compiler warning for this unreachable line. The old code didn't.

**Review by gvanrossum — COMMENTED** [review]
Can you run the new files through Black?

File: Tools/cases_generator/lexer.py (hunk 5 above)
Comment (markshannon) [reply Tools/cases_generator/lexer.py]: I've restored the old regex.

File: Tools/cases_generator/cwriter.py (hunk 3 above)
Comment (markshannon) [reply Tools/cases_generator/cwriter.py]: I'm finding MyPy to be close to unusable.
It will not do the most basic inference. Having to annotate every function that does not have a `return` with `-> None` is just extra work and noise.

Comment (markshannon) [reply Tools/cases_generator/cwriter.py]: Better architecture is more of a help to understanding code, IMO. 
If MyPy is an obstacle to that, then we should drop it.
I'm not opposed to type annotations, but we need to use a better checker.

Comment (gvanrossum) [reply Tools/cases_generator/cwriter.py]: Let’s talk options when I am back.

File: Makefile.pre.in (hunk 11 above)
Comment (markshannon) [reply Makefile.pre.in:1590]: I've change `generate_cases.py` to no longer generate the tier 1 code.

File: Tools/cases_generator/cwriter.py (hunk 3 above)
Comment (markshannon) [reply Tools/cases_generator/cwriter.py]: I've worked around the MyPy issues. Mainly renaming variables, adding `__all__`, and adding a bunch of ` -> None`s.

Not being able to trust the output of MyPy undermines the value of the one or two errors it did find.

File: Tools/cases_generator/analyzer.py [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+
+    @property
+    def size(self) -> int:
+        return 1 + sum(uop.size for uop in self.uops)
+
+
+@dataclass
+class PseudoInstruction:
+    name: str
+    targets: list[Instruction]
```
Comment (iritkatriel) [inline Tools/cases_generator/analyzer.py:159]: Remember to add the flags (my PR from yesterday)

File: Tools/cases_generator/parser.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    AstNode,
+)
+from formatting import prettify_filename
+
+
+BEGIN_MARKER = "// BEGIN BYTECODES //"
+END_MARKER = "// END BYTECODES //"
+
+
+Context = Context
```
Comment (iritkatriel) [inline Tools/cases_generator/parser.py]: why is this needed?

File: Tools/cases_generator/analyzer.py (hunk 9 above)
Comment (iritkatriel) [reply Tools/cases_generator/analyzer.py:39]: And a ``__str__`` (so `dump()` might not be needed))

File: Tools/cases_generator/analyzer.py [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+    type: str | None
+    condition: str | None
+    size: str
+    peek: bool = False
+
+    def __str__(self) -> str:
+        cond = " if (" + self.condition + ")" if self.condition else ""
+        size = "[" + self.size + "]" if self.size != "1" else ""
+        type = "" if self.type is None else self.type + " "
+        return f"{type}{self.name}{size}{cond} {self.peek}"
```
Comment (iritkatriel) [inline Tools/cases_generator/analyzer.py:78]: It would probably be nicer to consistently use f-strings.

File: Tools/cases_generator/analyzer.py [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+) -> SyntaxError:
+    return analysis_error(
+        f"Duplicate definition of '{name}' @ {context} "
+        f"previous definition @ {prev_context}",
+        token,
+    )
+
+
+def convert_stack_item(item: parser.StackEffect) -> StackItem:
+    return StackItem(item.name, item.type, item.cond, (item.size if item.size else "1"))
```
Comment (iritkatriel) [inline Tools/cases_generator/analyzer.py]:
```suggestion
    return StackItem(item.name, item.type, item.cond, (item.size or "1"))
```

File: Tools/cases_generator/analyzer.py [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+        return Properties(
+            any(part.properties.escapes for part in self.uops),
+            all(part.properties.infallible for part in self.uops),
+            any(part.properties.deopts for part in self.uops),
+            any(part.properties.oparg for part in self.uops),
+            any(part.properties.jumps for part in self.uops),
+            any(part.properties.ends_with_eval_breaker for part in self.uops),
+            any(part.properties.needs_this for part in self.uops),
+            any(part.properties.always_exits for part in self.uops),
+            any(part.properties.stores_sp for part in self.uops),
```
Comment (iritkatriel) [inline Tools/cases_generator/analyzer.py]: The knowledge of which fields exist in properties belong in the Properties class rather than here. You could define a function there that takes a list of Property objects and does this.

File: Tools/cases_generator/analyzer.py [hunk 18]
```diff
@@ -0,0 +1,470 @@
...
+    return Properties(
+        makes_escaping_api_call(op),
+        is_infallible(op),
+        variable_used(op, "DEOPT_IF"),
+        variable_used(op, "oparg"),
+        variable_used(op, "JUMPBY"),
+        eval_breaker_at_end(op),
+        variable_used(op, "this_instr"),
+        always_exits(op),
+        variable_used(op, "STORE_SP"),
```
Comment (iritkatriel) [inline Tools/cases_generator/analyzer.py]: I think it would be easier to maintain if these are passed as kwargs.

File: Tools/cases_generator/parser.py (hunk 14 above)
Comment (markshannon) [reply Tools/cases_generator/parser.py]: I was trying to convince MyPy not to complain. Apparently the way to do that is to define `__all__`.
I'll remove it.

https://github.com/python/mypy/issues/16596

File: Tools/cases_generator/analyzer.py (hunk 9 above)
Comment (markshannon) [reply Tools/cases_generator/analyzer.py:39]: `dump` handles indentation, so it's worth keeping.

File: Tools/cases_generator/analyzer.py [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+        self.properties.dump("    " + indent)
+
+    @property
+    def size(self) -> int:
+        if self._size < 0:
+            self._size = sum(c.size for c in self.caches)
+        return self._size
+
+
+Part = Uop | Skip
```
Comment (iritkatriel) [inline Tools/cases_generator/analyzer.py:127]: `Skip` is "unused cache entry", but here it's used as a uop?

File: Tools/cases_generator/stack.py [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+        popped = sorted(self.popped)
+        pushed = sorted(self.pushed)
+        self.popped = []
+        self.pushed = []
+        while popped and pushed:
+            pop = popped.pop()
+            push = pushed.pop()
+            if pop == push:
+                pass
+            elif pop > push:
```
Comment (iritkatriel) [inline Tools/cases_generator/stack.py:45]: pop and push are string representations of the size, right? What is this comparison doing?

File: Tools/cases_generator/analyzer.py (hunk 19 above)
Comment (markshannon) [reply Tools/cases_generator/analyzer.py:127]: Where is `Part` used as a `Uop`?

File: Tools/cases_generator/stack.py (hunk 20 above)
Comment (markshannon) [reply Tools/cases_generator/stack.py:45]: Probably not what I meant it to be doing.

Comment (markshannon) [reply Tools/cases_generator/stack.py:45]: Actually it is doing what it is supposed to.
I'll add a comment.

File: Tools/cases_generator/analyzer.py (hunk 19 above)
Comment (iritkatriel) [reply Tools/cases_generator/analyzer.py:127]: `uops: list[Part]` in line 133

File: Tools/cases_generator/mypy.ini [hunk 21]
```diff
@@ -11,3 +11,5 @@ strict = True
 strict_concatenate = True
 enable_error_code = ignore-without-code,redundant-expr,truthy-bool,possibly-undefined
 warn_unreachable = True
+allow_redefinition = True
+implicit_reexport = True
```
Comment (AlexWaygood) [inline Tools/cases_generator/mypy.ini:15]: Suggestion: move these lines of the config file above the `# ...And be strict:` comment on line 9, since they don't relate to increasing the strictness of mypy

File: Tools/cases_generator/analyzer.py [hunk 22]
```diff
@@ -0,0 +1,458 @@
+from dataclasses import dataclass
+import lexer
+import parser
+from typing import Optional
+
+
+@dataclass
```
Comment (AlexWaygood) [inline Tools/cases_generator/analyzer.py:7]: I would be tempted to do this, for readability. I agree with Guido's opinion that it's hard to read when instances of this class are constructed using positional arguments

```suggestion
@dataclass(kw_only=True)
```

File: Tools/cases_generator/analyzer.py [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+
+EXITS = set(
+    [
+        "DISPATCH",
+        "GO_TO_INSTRUCTION",
+        "Py_UNREACHABLE",
+        "DISPATCH_INLINED",
+        "DISPATCH_GOTO",
+    ]
+)
```
Comment (AlexWaygood) [inline Tools/cases_generator/analyzer.py]: Just use a literal set?

```suggestion
EXITS = {
    "DISPATCH",
    "GO_TO_INSTRUCTION",
    "Py_UNREACHABLE",
    "DISPATCH_INLINED",
    "DISPATCH_GOTO",
}
```

File: Tools/cases_generator/tier1_generator.py [hunk 24]
```diff
@@ -0,0 +1,417 @@
...
+from lexer import Token
+from stack import StackOffset
+
+
+HERE = os.path.dirname(__file__)
+ROOT = os.path.join(HERE, "../..")
+THIS = os.path.relpath(__file__, ROOT).replace(os.path.sep, "/")
+
+DEFAULT_INPUT = os.path.relpath(os.path.join(ROOT, "Python/bytecodes.c"))
+DEFAULT_OUTPUT = os.path.relpath(os.path.join(ROOT, "Python/generated_cases.c.h"))
```
Comment (AlexWaygood) [inline Tools/cases_generator/tier1_generator.py:31]: pathlib might make this slightly more elegant:

```py
from pathlib import Path

HERE = Path(__file__).parent
ROOT = HERE.parent.parent
THIS = Path(__file__).relative_to(ROOT).as_posix()

DEFAULT_INPUT = (ROOT / "Python/bytecodes.c").absolute()
DEFAULT_OUTPUT = (ROOT / "Python/generated_cases.c.h").absolute()
```

**Review by AlexWaygood — COMMENTED** [review]
Some optional suggestions:

File: Tools/cases_generator/stack.py [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+        # Special case simplification
+        if var.condition == "oparg & 1" and var.size == "1":
+            return f"({var.condition})"
+        else:
+            return f"(({var.condition}) ? {var.size} : 0)"
+    else:
+        return var.size
+
+
+@dataclass
```
Comment (AlexWaygood) [inline Tools/cases_generator/stack.py]: This dataclass appears to have 0 fields, since there are no annotations in the class body. Does it need to be a dataclass at all?

```suggestion
```

File: Tools/cases_generator/tier1_generator.py [hunk 26]
```diff
@@ -0,0 +1,417 @@
...
+
+class SizeMismatch(Exception):
+    pass
+
+
+class Stack:
+    def __init__(self) -> None:
+        self.top_offset: StackOffset = StackOffset()
+        self.base_offset: StackOffset = StackOffset()
+        self.peek_offset: StackOffset = StackOffset()
```
Comment (AlexWaygood) [inline Tools/cases_generator/tier1_generator.py]: These annotations feel somewhat redundant, and mypy still passes if you get rid of them:

```suggestion
        self.top_offset = StackOffset()
        self.base_offset = StackOffset()
        self.peek_offset = StackOffset()
```

File: Tools/cases_generator/analyzer.py (hunk 19 above)
Comment (markshannon) [reply Tools/cases_generator/analyzer.py:127]: It is a list of parts, and used as such. I'll change the name.

File: Tools/cases_generator/tier1_generator.py (hunk 24 above)
Comment (markshannon) [reply Tools/cases_generator/tier1_generator.py:31]: That is better, but I think there are enough changes in this PR for now.
I'll need to refactor this code to split off the other generators, so I can use pathlib then.

File: Tools/cases_generator/cwriter.py (hunk 3 above)
Comment (markshannon) [reply Tools/cases_generator/cwriter.py]: With the correct (but non-obvious) flags set, it works as expected.
